openssl/src/pkcs7.rs

```diff
@@ -4,15 +4,102 @@ use libc::c_int;
 
+use crate::asn1::{Asn1IntegerRef, Asn1StringRef};
 use crate::bio::{MemBio, MemBioSlice};
 use crate::error::ErrorStack;
 use crate::pkey::{HasPrivate, PKeyRef};
-use crate::stack::{Stack, StackRef};
+use crate::stack::{Stack, StackRef, Stackable};
 use crate::symm::Cipher;
+use crate::util::ForeignTypeRefExt;
 use crate::x509::store::X509StoreRef;
-use crate::x509::{X509Ref, X509};
+use crate::x509::{X509AlgorithmRef, X509NameRef, X509Ref, X509};
 use crate::{cvt, cvt_p};
 
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::PKCS7_SIGNER_INFO;
+    fn drop = ffi::PKCS7_SIGNER_INFO_free;
+
+    /// A PKCS#7 SignerInfo structure.
+    pub struct Pkcs7SignerInfo;
+
+    /// Reference to `Pkcs7SignerInfo`
+    pub struct Pkcs7SignerInfoRef;
+}
+
+impl Pkcs7SignerInfoRef {
+    /// Returns the issuer's subject name.
+    ///
+    /// This corresponds to `PKCS7_SIGNER_INFO`'s `issuer_and_serial.issuer` field.`
+    pub fn subject_name(&self) -> &X509NameRef {
+        unsafe {
+            let ias = (*self.as_ptr()).issuer_and_serial;
+            assert!(!ias.is_null());
+
+            let issuer = (*ias).issuer;
+            X509NameRef::from_const_ptr_opt(issuer).expect("subject name must not be null")
+        }
+    }
+
+    /// Returns the issuer's serial number.
+    ///
+    /// This corresponds to `PKCS7_SIGNER_INFO`'s `issuer_and_serial.serial` field.
+    pub fn serial_number(&self) -> &Asn1IntegerRef {
+        unsafe {
+            let ias = (*self.as_ptr()).issuer_and_serial;
+            assert!(!ias.is_null());
+
+            let serial = (*ias).serial;
+            Asn1IntegerRef::from_const_ptr_opt(serial).expect("serial number must not be null")
+        }
+    }
+
+    /// Returns the signature's digest algorithm.
+    pub fn digest_algorithm(&self) -> &X509AlgorithmRef {
+        unsafe {
+            let mut algor = ptr::null_mut();
+            ffi::PKCS7_SIGNER_INFO_get0_algs(
+                self.as_ptr(),
+                ptr::null_mut(),
+                &mut algor,
+                ptr::null_mut(),
+            );
+
+            X509AlgorithmRef::from_const_ptr_opt(algor).expect("digest algorithm must not be null")
+        }
+    }
+
+    /// Returns the signature's digest encryption algorithm.
+    pub fn digest_encryption_algorithm(&self) -> &X509AlgorithmRef {
+        unsafe {
+            let mut algor = ptr::null_mut();
+            ffi::PKCS7_SIGNER_INFO_get0_algs(
+                self.as_ptr(),
+                ptr::null_mut(),
+                ptr::null_mut(),
+                &mut algor,
+            );
+
+            X509AlgorithmRef::from_const_ptr_opt(algor)
+                .expect("digest encryption algorithm must not be null")
+        }
+    }
+
+    /// Returns the raw signature.
+    ///
+    /// This corresponds to `PKCS7_SIGNER_INFO`'s `enc_digest` field.
+    pub fn signature(&self) -> &Asn1StringRef {
+        unsafe {
+            // ASN1_OCTET_STRING is a typedef of ASN1_STRING
```

Comment:
They seem to be distinct types as far as I can tell: https://github.com/openssl/openssl/blob/master/include/openssl/asn1.h.in

Reply:
Hmm, this is what I was looking at: https://github.com/openssl/openssl/blob/6d2e0076e6a8146f3bdaa60151b2fa6e4df0d1bf/include/openssl/types.h#L48

Both `ASN1_OCTET_STRING` and `ASN1_STRING` are typedefs of `struct asn1_string_st`.

Reply:
Sure, but they have distinct APIs in the asn1.h header, and we currently define separate types mapping to many of those already: https://docs.rs/openssl/0.10.34/openssl/asn1/index.html

Reply:
We should add `Asn1OctetString`. Suggestion:  As this PR is not automatically mergeable any more due to changes in the directory structure (-> handwritten/...) and as I will need signer info from PKCS7, too, I can integrate the changes in my [PR](https://github.com/sfackler/rust-openssl/pull/1598). @chenxiaolong @sfackler: OK?

Reply:
Yes, please. Thank you!